valhalla/filesystem.h

```diff
@@ -416,22 +422,53 @@ inline void resize_file(const path& p, std::uintmax_t new_size) {
 }
 
 inline bool remove(const path& p) {
-  return ::remove(p.c_str()) == 0;
+  bool ret = ::remove(p.c_str()) == 0;
+
+  // it can occur that another thread is removing the same file
+  // object at the same time. Occasionally it will occur that
+  // while one thread is deleting (thread 1), the other thread
+  // (thread 2) will receive -1 from ::remove() and (errno==ENOENT
+  // or errno==EINVAL). And yet, the file still exists until thread 1
+  // is finished deleting. In this exact moment, if thread 2 calls
+  // stat() on the file object in question it will get 0 which means
+  // "file object exists". We kinda want thread 2 to wait for the
+  // file object to truly go out of existence. Hence, this spin loop.
```

Comment:
while i get that this is a nice feature i really dont think it should be the filesystem libraries concern if you write code that is racing to delete the same paths. that synchronization should be handled in the calling application. indeed i had a look at both the boost implementation as well as another filesystem implementation (https://github.com/gulrak/filesystem) and neither of them implement such a tactic: 
https://github.com/gulrak/filesystem/blob/de57485877865156370f5e9878592293645b207e/include/ghc/filesystem.hpp#L4396
https://github.com/boostorg/filesystem/blob/develop/src/operations.cpp#L229

@ktatterso what was the issue that triggered you looking into this anyway? i've kind of lost the plot at this point :smile:  i cant think of any multithreaded calls to `remove`. in mjolnir at the beginning of the process we `remove_all` existing tile directories in the main thread... what am i missing here?

Reply:
yeah it seems like its not something we really need to worry about at this time. the only call was the one that i remembered off the top of my head, the rest are all in tests which definitely shouldnt race because they definitely shouldnt be using the same tile directories:

![image](https://user-images.githubusercontent.com/697548/106910525-4c578680-66cf-11eb-968a-3d7139dbeac0.png)

now i get that you could be racing at the process level but again should filesystem really be the one protecting against that? i guess since it has such a small impact its probably ok but its again likely a deviation from the standard implimentation. i'm ok with whichever direction you want to go. want to just ship what we've got?

Reply:
Multithreaded folder creation motivated the creation of the unit test.  Deciding to test multithreaded directory removal was simply my idea.  I figured if we want to be robust in multithreaded directory creation we would also want to be robust in multithreaded directory removal.  In a sense, valhalla's filesystem::remove_all() is the "calling application" - so imo we should take responsibility for it being robust.

In my mind, I wrote a test that should work (threads concurrently calling filesystem::remove_all()).  It exposed an issue and I've resolved the issue.  No need to wait for a customer to find this issue, we can fix it now.